### PR TITLE
Add class fields and private methods

### LINKED ISSUE #154: Private Class Field representation
Some discussion in https://github.com/babel/babylon/pull/260 currently, figured I'd raise it over here before we accept anything.

Current proposal there being
```
interface PrivateName <: Expression, Pattern {
  type: "PrivateName";
  name: string;
}

interface ClassPrivateProperty <: Node {
  type: "ClassPrivateProperty";
  key: Identifier;
  value: Expression;
}

extend interface ClassBody {
  body: [ MethodDefinition | ClassPrivateProperty ]
}
```

I'm not sure I love the duplication between `PrivateName` and `Identifier` so perhaps `name` could be an `Identifier` instead of a string.

Keep in mind that both `this.#priv = 1234` and `#priv = 1234` are allowed, otherwise there'd be the potential for `MemberExpression` to just have a `private: boolean` property.

### PR DESCRIPTION
**Pull request by adrianheine**
Closes #154.

### CODE REVIEW

File: experimental/class-fields.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+interface PrivateName <: Node {
+    type: "PrivateName";
+    name: string;
+}
+
+interface FieldDefinition <: Node {
+    type: "FieldDefinition";
+    key: PrivateName | Expression;
+    value: Expression | null;
+    computed: boolean;
```
[the remark this answers is not in the document]
Comment (adrianheine) [reply experimental/class-fields.md]: I would expect so, `type` is also a literal value. Are you thinking about that?

```js
interface PrivateFieldDefinition {
  key: PrivateName;
  computed: false;
}
```

[the remark this answers is not in the document]
Comment (j-f1) [reply experimental/class-fields.md]: Yep, that’s exactly what I was thinking of.

[the remark this answers is not in the document]
Comment (adrianheine) [reply experimental/class-fields.md]: We could introduce a new interface `ClassElement` with `key`, `value` and `computed` and make `MethodDefinition`, `PublicFieldDefinition` and `PrivateFieldDefinition` extend it.

[the remark this answers is not in the document]
Comment (danez) [reply experimental/class-fields.md]: I think that would make sense to create an interface for ClassElement. But I wonder if it makes more sense to not include `computed` in this new interface and completely leave it out for PrivateFields?

[the remark this answers is not in the document]
Comment (RReverser) [reply experimental/class-fields.md]: Yeah, if `computed` is always `false`, and this node doesn't inherit anything, I think it's better to remove that property altogether.

**Comment by adrianheine** [comment]
How about this:

```js
interface ClassBody <: Node {
    type: "ClassBody";
    body: [ ClassElement ];
}

abstract interface ClassElement <: Node {
    static: boolean;
    computed: boolean;
    key: Expression | PrivateName;
    value: Expression | null;
}

interface MethodDefinition <: ClassElement {
    type: "MethodDefinition";
    value: FunctionExpression;
    kind: "constructor" | "method" | "get" | "set";
}

interface FieldDefinition <: ClassElement {
    type: "FieldDefinition";
}
```
Some thoughts:
* Whether a `ClassElement` is private can only be determined by the type of the `key` (there is no redundancy)
* For private `ClassElement`s, `computed` is always `false` (but this fact is not represented)
* For `MethodDefinition`s with `kind` `"constructor"`, `computed` and `static` are always `false` (but this fact is not represented)
* Adding [static blocks](https://tc39.github.io/proposal-class-static-block/#sec-class-definitions) will not be straightforward

**Comment by piranna** [comment]
Can this be merged? It's blocking support of private methods in babel and babel-eslint...

**Comment by adrianheine** [comment]
@piranna Do the emitted AST follow this proposal?

**Comment by nicolo-ribaudo** [comment]
Not in Babel, only babel-eslint follows the ESTree spec.

**Comment by piranna** [comment]
> @piranna Does the emitted AST follow this proposal?

I don't know, I just only use that tools, and found a bug that seems to be directly related to this pull-request to being fixed... :-/

**Comment by filmic** [comment]
@piranna Babel transpiles code with private methods without issues if you use `@babel/plugin-proposal-private-methods`.

This PR will eventually solve issue with ESLint failing to traverse code with private methods when using `babel-eslint` parser.

**Comment by piranna** [comment]
> @piranna Babel transpiles code with private methods without issues if you use `@babel/plugin-proposal-private-methods`.

I'm already using this plugin without problems... Fair enough, the problem only affect `babel-eslint`.

> This PR will eventually solve issue with ESLint failing to traverse code with private methods when using `babel-eslint` parser.

Ok, thank you for the clarification :+1:

**Comment by JounQin** [comment]
Any update?

**Comment by anthonator** [comment]
Anything the community can do to help push this along?

**Comment by RReverser** [comment]
@adrianheine is this up-to-date with what the tools are doing? If so, feel free to merge, especially since it goes into experimental folder anyway.

**Comment by armano2** [comment]
Do we have a consensus on how AST structure should look for this, namely private Identifier?

**Comment by mysticatea** [comment]
I think that `private:boolean` flag (or `visibility:string` field) is better than `PrivateName` node.

In semantics, the AST should represent private fields, private methods, and private member accesses. But private names don't exist. I feel that the current proposal is tied to source code text seriously. It doesn't look AST.

**Comment by bradzacher** [comment]
This is currently blocking us (typescript-eslint) from adding support for typescript 3.8.

Is there something we can do to help this progress?

File: experimental/private-methods.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    type: "PrivateName";
+    name: string;
+}
+
+extend interface MethodDefinition {
+    key: Expression | PrivateName;
+}
+```
+
+Private methods cannot be static or have computed names.
```
Comment (JLHwung) [inline experimental/private-methods.md:14]: Per https://github.com/tc39/proposal-static-class-features/, private methods can be static.

File: experimental/private-methods.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+```js
+interface PrivateMemberExpression <: Expression, Pattern {
```
Comment (JLHwung) [inline experimental/private-methods.md:17]: Recently the [support](https://tc39.es/proposal-class-fields/#prod-OptionalChain) of optional chaining has been added to this proposal, and both `?. PrivateIdentifier` and `OptionalChain . PrivateIdentifier` have been accepted. If we are to be consistent with current [`ChainingExpression`](https://github.com/estree/estree/blob/a75c6933bcad86174d2c913eea6d1cf87db234f7/es2020.md#chainingexpression) approach in #204, does it mean that we should also add `PrivateMemberChain`?
```js
interface PrivateMemberChain extends Chain {
  type: "PrivateMemberChain";
  property: PrivateName
}
```

Personally I lean to merge `PrivateMemberChain` with `MemberChain`
```js
interface MemberChain extends Chain {
  type: "MemberChain"
  computed: boolean
  property: Expression
}
```
with an additional note that if property is a `PrivateName`, `computed` must be `false`.

**Comment by JLHwung** [comment]
> In semantics, the AST should represent private fields, private methods, and private member accesses. But private names don't exist.

Note that in #217 a private names is allowed in the `left` of a `BinaryExpression[operator=in]`.

**Comment by mysticatea** [comment]
> Note that in #217 a private names is allowed in the left of a BinaryExpression[operator=in].

Thank you for teaching. In that case, I'm fine with the `PrivateName` node.

**Comment by nzakas** [comment]
@JLHwung does this proposal reflect what Babel is currently doing? Or should we close in favor of a spec from you?

**Comment by JLHwung** [comment]
@nzakas Babel splits both `FieldDefinition ` and `MethodDefinition` into `ClassProperty/ClassMethod` and their private versions.

Besides that we don't have `PrivateMemberExpression` but extend the current `MemberExpression` with `PrivateName` support.

I can draft a new PR based on Babel's approach. I don't have strong opinion on closing this one since these two are easily interchangeable.

**Comment by nzakas** [comment]
Cool, thanks. Should we close this PR then?

**Comment by RReverser** [comment]
@nzakas Given that discussion here sort of depends on decision on optional chaining, I'd say let's leave this open. This PR looks quite reasonable as well, but it's hard to tell which one is better at this point.

**Comment by nzakas** [comment]
I guess my point is that since this is in the experimental folder, and we're only giving a cursory look at experimental syntax, what's the expectation when two competing specs are submitted?

**Pull request closed without merging**